rust/rope/src/diff.rs

```diff
@@ -245,21 +271,82 @@ impl DiffBuilder {
     }
 }
 
+/// An interval for a specific rope.
+#[derive(Debug, Clone)]
+pub struct RopeSlice<'a> {
+    rope: &'a Rope,
+    range: Range<usize>,
+}
+
+impl PartialEq for RopeSlice<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        let mut it1 =
+            self.rope.iter_chunks(Interval::from(self.range.clone())).flat_map(|x| x.chars());
+        let mut it2 =
+            other.rope.iter_chunks(Interval::from(other.range.clone())).flat_map(|x| x.chars());
+
+        loop {
+            let c1 = it1.next();
+            let c2 = it2.next();
+
+            if c1 != c2 {
+                return false;
+            }
+
+            if c1.is_none() {
+                return true;
+            }
+        }
+    }
+}
+
+impl Eq for RopeSlice<'_> {}
+
+impl Hash for RopeSlice<'_> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let iter =
+            self.rope.iter_chunks(Interval::from(self.range.clone())).flat_map(|x| x.chars());
```

Comment:
ditto chars being slow here.

It might be worth looking into what the contract is around the hash impls of `std` types; I suspect we can actually just hash the chunks returned by `iter_chunks` one by one and get the same hash for a given string, regardless of how it is broken up.